### PR TITLE
Fix `pm.distributions.transforms.ordered` fails on >1 dimension

### PR DESCRIPTION
**Pull request by purna135**
Addressing #5659

Kept the original dims in the log_jac_det by changing
```
def log_jac_det(self, value, *inputs): 
     return at.sum(value[..., 1:], axis=-1)
```
to 
```
def log_jac_det(self, value, *inputs):
    return at.sum(value[..., 1:], axis=-1, keepdims=True)
```

### CODE REVIEW

**Comment by purna135** [comment]
Hello @ricardoV94, after making these changes I ran the code mentioned in the issue, and it ran without error this time.
However, I am unsure how to select the appropriate test for this; could you please assist me with this?

**Comment by ricardoV94** [comment]
You can just add the original issue example as a new test to `test_transforms.py` and assert the result of `point_logps` is what's expected.

**Comment by purna135** [comment]
test has been added, please have a look.

File: pymc/tests/test_transforms.py
Comment (ricardoV94) [inline pymc/tests/test_transforms.py]: We don't need to use coords for this example.

```suggestion
def test_transforms_ordered():

    with pm.Model() as model:
        kappa = pm.Normal(
            "kappa",
            mu=[-3, -1, 1, 2],
            sigma=1,
            size=(10, 4),
            transform=pm.distributions.transforms.ordered,
        )
```

Also should test with the other transform that you have changed.

**Review by ricardoV94 — CHANGES_REQUESTED** [review]
Looks good. Left a suggestion and request

File: pymc/tests/test_transforms.py
Comment (purna135) [reply pymc/tests/test_transforms.py]: Done : )
but what about the other tests which are falling ?

**Comment by drbenvincent** [comment]
Thanks for the contribution so far @purna135 ! I have my eye on this, and it looks like there are just two failing tests before this can be merged?

**Comment by purna135** [comment]
Thank you so much, @drbenvincent; I was hoping for a response.
Yes, some tests are failing, but I can't figure out how to fix them; could you please assist me?

**Comment by drbenvincent** [comment]
> Thank you so much, @drbenvincent; I was hoping for a response.
> Yes, some tests are failing, but I can't figure out how to fix them; could you please assist me?

I would gladly, if I could, but so far my contributions have been to `pymc-examples`, and not to `pymc` itself. So I do not have enough experience. Maybe @ricardoV94 could give some hints when he has time?

**Comment by ricardoV94** [comment]
@purna135 Have you tried to investigate why those tests are failing locally?

**Comment by purna135** [comment]
Yes, I ran the test locally, but I couldn't figure out what was causing it.
I assumed that because we changed `log_jac_det`, the expectation of the following line changed as well. 
`jacob_det = transform.log_jac_det(test_array_transf, *x.owner.inputs)`

As a result, I think we should revise the `assert` statement.

**Comment by ricardoV94** [comment]
You can print the results before and after the changes to be sure what's changed. Then we have to figure out if we introduced a bug or if the old results were wrong (probably the former)

**Comment by purna135** [comment]
Is this `jacob_det`  the one I should print to check the results?

**Comment by purna135** [comment]
Hi @ricardoV94 I'm still waiting for a response and am unsure how to proceed. Could you please assist me?

File: pymc/distributions/transforms.py
Comment (Sayam753) [inline pymc/distributions/transforms.py:85]: Setting `keepdims=True` will keep the last dimension while computing logprob for multivariate distributions. This behaviour can be a lot confusing to interpret when adding transforms for multivariate distributions.

Example:
```python
from pymc.distributions.logprob import joint_logpt

with pm.Model() as model:
    mv = pm.MvNormal(
        "mv",
        mu=[-3, -1, 1, 2],
        cov=np.eye(4),
        initval=np.random.randn(4),
    )

print(joint_logpt(mv, sum=False)[0].ndim)  # will output 0
```

After adding Ordered transform
```python
from pymc.distributions.logprob import joint_logpt

with pm.Model() as model:
    mv = pm.MvNormal(
        "mv",
        mu=[-3, -1, 1, 2],
        cov=np.eye(4),
        initval=np.random.randn(4),
        transform=pm.distributions.transforms.ordered,
    )

print(joint_logpt(mv, sum=False)[0].ndim)  # will output 1
```

I think there can be a simple check here: https://github.com/aesara-devs/aeppl/blob/65d25c39fa771809b4d992c9323f59e78f282cd5/aeppl/transforms.py#L372-L388

to squeeze out the last dimension in case when transforms are added for multivariate distributions.

cc @ricardoV94

Comment (ricardoV94) [reply pymc/distributions/transforms.py:85]: We might need to tell the transform what is the support dimensionality of the RV. Can be an internal variable like `n` in the transform that's used by the LKJCholeskyCov

Comment (Sayam753) [reply pymc/distributions/transforms.py:85]: > We might need to tell the transform what is the support dimensionality of the RV.

If I get it right, we need to somehow pass `op.ndim_supp` to transforms to decide whether or not to keep last dimension. Seems like a simpler solution. 

Also, I think [SimplexTransform defined in AePPL](https://github.com/aesara-devs/aeppl/blob/main/aeppl/transforms.py#L289) needs to account for this change.

Comment (ricardoV94) [reply pymc/distributions/transforms.py:85]: We don't need to pass it to all transforms, we can instead initialize this transform with that information. Like we initialize Interval or LKJ transform with the specific information they need. This would follow a similar API and be less disruptive

In this case instead of having a single ordered already initialized, the user would initialize it (or us behind the scenes, like pm.interval does). It can be a kwarg that by default is 0 (for univariate RVs)

pseudo-code:
```python
class OrderedTransform(TransformRV):

  def __init__(self, ndim_supp):
    self.ndim_supp = ndim_supp

  def log_jac_det(self, value, *inputs):
    # use self.ndim_supp to return the right shape

def ordered(ndim_supp=0):
    return OrderedTransform(ndim_supp)
```

Comment (ricardoV94) [reply pymc/distributions/transforms.py:85]: Is there anything wrong with the Simplex? That seems to be vector specific transform, I don't think it would be used in other cases.

We can think of it as being a VectorSimplex (i.e., a Simplex for RVs with ndim_supp=1), which is what it is used for in PyMC. If we wanted to use it with other ndim_supp we could make it more flexible...

Comment (Sayam753) [reply pymc/distributions/transforms.py:85]: Passing the required info to transforms during initialization indeed makes sense. @purna135 can you take up adding the proposed change in this PR itself?

> Is there anything wrong with the Simplex?

I noticed a test in `test_transforms.py` that applies Simplex transform to a batched univariate distribution. 

https://github.com/pymc-devs/pymc/blob/bae5087621323af3cb3cc95cf8d673da21ab4023/pymc/tests/test_transforms.py#L509-L526

This prompted me to make sure that even Simplex should not reduce the last dimension in case when applied to univariate distributions.

Comment (purna135) [reply pymc/distributions/transforms.py:85]: Thank you, @Sayam753. I'll work on these suggested changes and let you know if I need any assistance.

**Comment by drbenvincent** [comment]
Hi. Just dropping a message here to see if anyone is able to push this forward? I have a vested interest in this fix, but am not familiar enough to help contribute efficiently.

**Comment by TimOliverMaier** [comment]
Hey! @purna135, is there any news on this? I would also be interested in this fix. I tried to add the suggestions into the code, but I am still having failing tests and tbh a lack of understanding, as well. I would be happy to help here, but I can't accomplish it by myself, too.

**Comment by TimOliverMaier** [comment]
I've been advancing on this locally. There are two remaining issues, as far as I see:
1.  There are some remaining failing tests that are caused by the changing dimensionality of the jacobian determinant <- Some 
     Transforms keep the last dimension, others do not. Keeping the other transforms as they are will make some further cases 
     in `test_transform.check_vectortransform_elementwise_logp()` necessary.
2.  Sampling of ordered multivariate distributions fails with `ValueError` :
     `"array must not contain infs or NaNs\nApply node that caused the error: SolveTriangular..."` 
     However, this is also the case on the main branch for me. 
What is the best way to collaborate on this PR? @purna135

**Comment by purna135** [comment]
Hello, @TimOliverMaier. Thank you so much for addressing this issue and sorry for the delay in responding; I was preoccupied with other issues and didn't have time to solve this one.
It would be greatly appreciated if you could open a new PR from your local branch and I would close this one.

**Comment by TimOliverMaier** [comment]
Hey, I opened a new PR. #6255 It is still work in progress, though.

**Comment by ricardoV94** [comment]
Fixed in #6255

**Pull request closed without merging**
